### PR TITLE
2) Add basic interface that allows to issue Floquet calibration requests on cirq.google.Engine

### PR DESCRIPTION
**Pull request by mrwojtek**
This change adds a simple containers that represent Floquet calibration requests and responses.

It contains a logic necessary to encode and decode the API requests.

This is a part of a series of PRs which can be grouped into:
 - #1 (1) and #2 (2) that are the low-level PRs which allow to encode and decode the Floquet calibration requests.
 - #3 (3), #5 (5) and #6 (6) that are a simple example on how the Floquet calibration can be used to correct for single-qubit Z phases induced by the two-qubit interacting gates.
 - #4 (4) which is a simulator that allows to test and debug angle-based circuit corrections.

TODOs:
 - Missing tests and documentation
 - Better naming
 - A number of TODOs for follow-up PRs

### CODE REVIEW

File: cirq/google/calibration/phased_fsim.py [hunk 1]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
             Otherwise the current values are used.
         """
         return other.merge_with(self)
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationOptions:
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: Add docstring

File: cirq/google/calibration/phased_fsim.py [hunk 2]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+    characterize_theta: bool
+    characterize_zeta: bool
+    characterize_chi: bool
+    characterize_gamma: bool
+    characterize_phi: bool
+
+
+# TODO: Fix json serialization (the default one doesn't work with tuples as dictionary keys).
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: If you want to create a bug and set this to NOT_YET_SERIALIZABLE I can take this off your plate.

File: cirq/google/calibration/phased_fsim.py [hunk 3]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+# TODO: Add start and end calibration timestamp
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: I would just create issues for these and leave them for later.  Myself or someone else can help out with this.

File: cirq/google/calibration/phased_fsim.py [hunk 4]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+            return None
+
+
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimCalibrationRequest(abc.ABC):
+    gate: Gate  # Any gate which can be described by cirq.PhasedFSim
+    gate_set: SerializableGateSet
+    pairs: Tuple[Tuple[Qid, Qid]]
+
+    @abc.abstractmethod
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py:223]: Do we think we will have multiple implementations of this interface?

File: cirq/google/calibration/workflow.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    PhasedFSimCalibrationResult
+)
+from cirq.google.engine import Engine
+
+
+def run_characterizations(calibrations: List[PhasedFSimCalibrationRequest],
+                          engine: Engine,
+                          processor_id: str,
+                          handler_name: str
+                          ) -> List[PhasedFSimCalibrationResult]:
```
Comment (dstrain115) [inline cirq/google/calibration/workflow.py]: Add docstring and tests for functions in this file.

File: cirq/google/calibration/phased_fsim.py [hunk 6]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationRequest(PhasedFSimCalibrationRequest):
+    options: FloquetPhasedFSimCalibrationOptions
+
+    def to_calibration_layer(self, handler_name: str) -> CalibrationLayer:
+        circuit = Circuit([self.gate.on(*pair) for pair in self.pairs])
+        return CalibrationLayer(
+            calibration_type='floquet_phased_fsim_characterization',
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: Should we make 'floquet_phased_fsim_characterization' a constant somewhere?  It looks like we will need it for  run_characterizations as well.

File: cirq/google/calibration/phased_fsim.py [hunk 7]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+                'est_theta': self.options.characterize_theta,
+                'est_zeta': self.options.characterize_zeta,
+                'est_chi': self.options.characterize_chi,
+                'est_gamma': self.options.characterize_gamma,
+                'est_phi': self.options.characterize_phi,
+                'readout_corrections': True
+            }
+        )
+
+    def parse_result(self, result: CalibrationResult) -> PhasedFSimCalibrationResult:
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py:280]: add docstring and tests

File: cirq/google/calibration/phased_fsim.py [hunk 8]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+        decoded = collections.defaultdict(lambda: {})
+        for keys, values in result.metrics['angles'].items():
+            for key, value in zip(keys, values):
+                match = re.match(r'(\d+)_(.+)', key)
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: We probably should document the key string format somewhere.

File: cirq/google/calibration/phased_fsim.py [hunk 9]
```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
...
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationResult(PhasedFSimCalibrationResult):
+    options: FloquetPhasedFSimCalibrationOptions
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationRequest(PhasedFSimCalibrationRequest):
+    options: FloquetPhasedFSimCalibrationOptions
+
+    def to_calibration_layer(self, handler_name: str) -> CalibrationLayer:
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: add a test for this.

File: cirq/google/calibration/phased_fsim.py (hunk 4 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py:223]: I think so, for XEB the options will be very different. The requests are calibration-specific.

File: cirq/google/calibration/phased_fsim.py (hunk 6 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: It was my oversight to put this string here, thanks for pointing out. My plan is to remove "handler_name" argument and make it a constant as you're suggesting.

File: cirq/google/calibration/phased_fsim.py (hunk 8 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: I intend to change it, I would prefer to wait with that until it's stable.

File: cirq/google/calibration/phased_fsim.py (hunk 3 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: OK, I can do that later.

File: cirq/google/calibration/phased_fsim.py (hunk 9 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: Done

File: cirq/google/calibration/workflow.py (hunk 5 above)
Comment (mrwojtek) [reply cirq/google/calibration/workflow.py]: Done

File: cirq/google/calibration/phased_fsim.py (hunk 2 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: Removed serialization from this PR

File: cirq/google/calibration/phased_fsim.py (hunk 1 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: Done.

File: cirq/google/calibration/phased_fsim.py (hunk 6 above)
Comment (mrwojtek) [reply cirq/google/calibration/phased_fsim.py]: Done.

**Comment by mpharrigan** [comment]
xref https://github.com/quantumlib/Cirq/pull/3690

**Pull request merged**